File: src/Hedera.Hashgraph.Tests/TransferTransactionHooksTest.cs
// SPDX-License-Identifier: Apache-2.0
using Org.Assertj.Core.Api.Assertions;
using Org.Junit.Jupiter.Api;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Text;

namespace Com.Hedera.Hashgraph.Sdk
{
    public class TransferTransactionHooksTest
    {
        virtual void ShouldAddHbarTransferWithPreTxAllowanceHook()
        {
            var tx = new TransferTransaction();
            var accountId = new AccountId(0, 0, 1);
            var amount = Hbar.FromTinybars(1000);
            var hookCall = new FungibleHookCall(123, new EvmHookCall(new byte[] { 1, 2, 3 }, 100000), FungibleHookType.PRE_TX_ALLOWANCE_HOOK);
            var result = tx.AddHbarTransferWithHook(accountId, amount, hookCall);
            AssertThat(result).IsSameAs(tx);
            AssertThat(tx.GetHbarTransfers()).HasSize(1);
            Assert.Equal(tx.GetHbarTransfers()[accountId], amount);
        }

        virtual void ShouldAddHbarTransferWithPrePostTxAllowanceHook()
        {
            var tx = new TransferTransaction();
            var accountId = new AccountId(0, 0, 1);
            var amount = Hbar.FromTinybars(2000);
            var hookCall = new FungibleHookCall(456, new EvmHookCall(new byte[] { 4, 5, 6 }, 200000), FungibleHookType.PRE_POST_TX_ALLOWANCE_HOOK);
            var result = tx.AddHbarTransferWithHook(accountId, amount, hookCall);
            AssertThat(result).IsSameAs(tx);
            AssertThat(tx.GetHbarTransfers()).HasSize(1);
            Assert.Equal(tx.GetHbarTransfers()[accountId], amount);
        }

        virtual void ShouldThrowExceptionForNullAccountId()
        {
            var tx = new TransferTransaction();
            var amount = Hbar.FromTinybars(1000);
            var hookCall = new FungibleHookCall(123, new EvmHookCall(new byte[] { 1, 2, 3 }, 100000), FungibleHookType.PRE_TX_ALLOWANCE_HOOK);
            AssertThatThrownBy(() => tx.AddHbarTransferWithHook(null, amount, hookCall)).IsInstanceOf(typeof(NullReferenceException)).HasMessage("accountId cannot be null");
        }

        virtual void ShouldThrowExceptionForNullAmount()
        {
            var tx = new TransferTransaction();
            var accountId = new AccountId(0, 0, 1);
            var hookCall = new FungibleHookCall(123, new EvmHookCall(new byte[] { 1, 2, 3 }, 100000), FungibleHookType.PRE_TX_ALLOWANCE_HOOK);
            AssertThatThrownBy(() => tx.AddHbarTransferWithHook(accountId, null, hookCall)).IsInstanceOf(typeof(NullReferenceException)).HasMessage("amount cannot be null");
        }

        virtual void ShouldThrowExceptionForNullHookCall()
        {
            var tx = new TransferTransaction();
            var accountId = new AccountId(0, 0, 1);
            var amount = Hbar.FromTinybars(1000);
            AssertThatThrownBy(() => tx.AddHbarTransferWithHook(accountId, amount, null)).IsInstanceOf(typeof(NullReferenceException)).HasMessage("hookCall cannot be null");
        }

        virtual void ShouldNotAllowNullHookTypeRemovedByTypedAPI()
        {

            // No-op: hook type is encoded in FungibleHookCall now
            AssertThat(true).IsTrue();
        }

        virtual void ShouldUpdateExistingTransferWithHook()
        {
            var tx = new TransferTransaction();
            var accountId = new AccountId(0, 0, 1);
            var amount = Hbar.FromTinybars(1000);

            // First add a regular transfer
            tx.AddHbarTransfer(accountId, amount);

            // Then add a hook to the existing transfer
            var hookCall = new FungibleHookCall(123, new EvmHookCall(new byte[] { 1, 2, 3 }, 100000), FungibleHookType.PRE_TX_ALLOWANCE_HOOK);
            var result = tx.AddHbarTransferWithHook(accountId, amount, hookCall);
            AssertThat(result).IsSameAs(tx);
            AssertThat(tx.GetHbarTransfers()).HasSize(1);
            Assert.Equal(tx.GetHbarTransfers()[accountId], Hbar.FromTinybars(2000));
        }

        virtual void ShouldCreateNewTransferWhenExistingTransferHasHook()
        {
            var tx = new TransferTransaction();
            var accountId = new AccountId(0, 0, 1);
            var amount = Hbar.FromTinybars(1000);

            // First add a transfer with a hook
            var hookCall1 = new FungibleHookCall(123, new EvmHookCall(new byte[] { 1, 2, 3 }, 100000), FungibleHookType.PRE_TX_ALLOWANCE_HOOK);
            tx.AddHbarTransferWithHook(accountId, amount, hookCall1);

            // Try to add another hook - should create a new transfer
            var hookCall2 = new FungibleHookCall(456, new EvmHookCall(new byte[] { 4, 5, 6 }, 200000), FungibleHookType.PRE_POST_TX_ALLOWANCE_HOOK);
            tx.AddHbarTransferWithHook(accountId, amount, hookCall2);
            AssertThat(tx.GetHbarTransfers()).HasSize(1);
            Assert.Equal(tx.GetHbarTransfers()[accountId], Hbar.FromTinybars(2000));
        }

        virtual void ShouldHandleMultipleAccountsWithHooks()
        {
            var tx = new TransferTransaction();
            var accountId1 = new AccountId(0, 0, 1);
            var accountId2 = new AccountId(0, 0, 2);
            var amount = Hbar.FromTinybars(1000);
            var hookCall1 = new FungibleHookCall(123, new EvmHookCall(new byte[] { 1, 2, 3 }, 100000), FungibleHookType.PRE_TX_ALLOWANCE_HOOK);
            var hookCall2 = new FungibleHookCall(456, new EvmHookCall(new byte[] { 4, 5, 6 }, 200000), FungibleHookType.PRE_POST_TX_ALLOWANCE_HOOK);
            tx.AddHbarTransferWithHook(accountId1, amount, hookCall1);
            tx.AddHbarTransferWithHook(accountId2, amount, hookCall2);
            AssertThat(tx.GetHbarTransfers()).HasSize(2);
            Assert.Equal(tx.GetHbarTransfers()[accountId1], amount);
            Assert.Equal(tx.GetHbarTransfers()[accountId2], amount);
        }

        virtual void ShouldThrowExceptionWhenFrozen()
        {
            var tx = new TransferTransaction();
            var accountId = new AccountId(0, 0, 1);
            var amount = Hbar.FromTinybars(1000);
            var hookCall = new FungibleHookCall(123, new EvmHookCall(new byte[] { 1, 2, 3 }, 100000), FungibleHookType.PRE_TX_ALLOWANCE_HOOK);

            // Set up the transaction properly before freezing
            tx.SetTransactionId(TransactionId.WithValidStart(AccountId.FromString("0.0.5006"), java.time.Instant.Now()));
            tx.SetNodeAccountIds(java.util.Arrays.AsList(AccountId.FromString("0.0.5005")));

            // Freeze the transaction
            tx.Freeze();
            AssertThatThrownBy(() => tx.AddHbarTransferWithHook(accountId, amount, hookCall)).IsInstanceOf(typeof(InvalidOperationException));
        }
    }
}

File: src/test/java/com/hedera/hashgraph/sdk/TransferTransactionHooksTest.java
// SPDX-License-Identifier: Apache-2.0
package com.hedera.hashgraph.sdk;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import org.junit.jupiter.api.Test;

public class TransferTransactionHooksTest {

    @Test
    void shouldAddHbarTransferWithPreTxAllowanceHook() {
        var tx = new TransferTransaction();
        var accountId = new AccountId(0, 0, 1);
        var amount = Hbar.fromTinybars(1000);
        var hookCall = new FungibleHookCall(
                123L, new EvmHookCall(new byte[] {1, 2, 3}, 100000L), FungibleHookType.PRE_TX_ALLOWANCE_HOOK);

        var result = tx.addHbarTransferWithHook(accountId, amount, hookCall);

        assertThat(result).isSameAs(tx);
        assertThat(tx.getHbarTransfers()).hasSize(1);
        assertThat(tx.getHbarTransfers().get(accountId)).isEqualTo(amount);
    }

    @Test
    void shouldAddHbarTransferWithPrePostTxAllowanceHook() {
        var tx = new TransferTransaction();
        var accountId = new AccountId(0, 0, 1);
        var amount = Hbar.fromTinybars(2000);
        var hookCall = new FungibleHookCall(
                456L, new EvmHookCall(new byte[] {4, 5, 6}, 200000L), FungibleHookType.PRE_POST_TX_ALLOWANCE_HOOK);

        var result = tx.addHbarTransferWithHook(accountId, amount, hookCall);

        assertThat(result).isSameAs(tx);
        assertThat(tx.getHbarTransfers()).hasSize(1);
        assertThat(tx.getHbarTransfers().get(accountId)).isEqualTo(amount);
    }

    @Test
    void shouldThrowExceptionForNullAccountId() {
        var tx = new TransferTransaction();
        var amount = Hbar.fromTinybars(1000);
        var hookCall = new FungibleHookCall(
                123L, new EvmHookCall(new byte[] {1, 2, 3}, 100000L), FungibleHookType.PRE_TX_ALLOWANCE_HOOK);

        assertThatThrownBy(() -> tx.addHbarTransferWithHook(null, amount, hookCall))
                .isInstanceOf(NullPointerException.class)
                .hasMessage("accountId cannot be null");
    }

    @Test
    void shouldThrowExceptionForNullAmount() {
        var tx = new TransferTransaction();
        var accountId = new AccountId(0, 0, 1);
        var hookCall = new FungibleHookCall(
                123L, new EvmHookCall(new byte[] {1, 2, 3}, 100000L), FungibleHookType.PRE_TX_ALLOWANCE_HOOK);

        assertThatThrownBy(() -> tx.addHbarTransferWithHook(accountId, null, hookCall))
                .isInstanceOf(NullPointerException.class)
                .hasMessage("amount cannot be null");
    }

    @Test
    void shouldThrowExceptionForNullHookCall() {
        var tx = new TransferTransaction();
        var accountId = new AccountId(0, 0, 1);
        var amount = Hbar.fromTinybars(1000);
        assertThatThrownBy(() -> tx.addHbarTransferWithHook(accountId, amount, null))
                .isInstanceOf(NullPointerException.class)
                .hasMessage("hookCall cannot be null");
    }

    @Test
    void shouldNotAllowNullHookTypeRemovedByTypedAPI() {
        // No-op: hook type is encoded in FungibleHookCall now
        assertThat(true).isTrue();
    }

    @Test
    void shouldUpdateExistingTransferWithHook() {
        var tx = new TransferTransaction();
        var accountId = new AccountId(0, 0, 1);
        var amount = Hbar.fromTinybars(1000);

        // First add a regular transfer
        tx.addHbarTransfer(accountId, amount);

        // Then add a hook to the existing transfer
        var hookCall = new FungibleHookCall(
                123L, new EvmHookCall(new byte[] {1, 2, 3}, 100000L), FungibleHookType.PRE_TX_ALLOWANCE_HOOK);

        var result = tx.addHbarTransferWithHook(accountId, amount, hookCall);

        assertThat(result).isSameAs(tx);
        assertThat(tx.getHbarTransfers()).hasSize(1);
        assertThat(tx.getHbarTransfers().get(accountId)).isEqualTo(Hbar.fromTinybars(2000));
    }

    @Test
    void shouldCreateNewTransferWhenExistingTransferHasHook() {
        var tx = new TransferTransaction();
        var accountId = new AccountId(0, 0, 1);
        var amount = Hbar.fromTinybars(1000);

        // First add a transfer with a hook
        var hookCall1 = new FungibleHookCall(
                123L, new EvmHookCall(new byte[] {1, 2, 3}, 100000L), FungibleHookType.PRE_TX_ALLOWANCE_HOOK);
        tx.addHbarTransferWithHook(accountId, amount, hookCall1);

        // Try to add another hook - should create a new transfer
        var hookCall2 = new FungibleHookCall(
                456L, new EvmHookCall(new byte[] {4, 5, 6}, 200000L), FungibleHookType.PRE_POST_TX_ALLOWANCE_HOOK);
        tx.addHbarTransferWithHook(accountId, amount, hookCall2);

        assertThat(tx.getHbarTransfers()).hasSize(1);
        assertThat(tx.getHbarTransfers().get(accountId)).isEqualTo(Hbar.fromTinybars(2000));
    }

    @Test
    void shouldHandleMultipleAccountsWithHooks() {
        var tx = new TransferTransaction();
        var accountId1 = new AccountId(0, 0, 1);
        var accountId2 = new AccountId(0, 0, 2);
        var amount = Hbar.fromTinybars(1000);

        var hookCall1 = new FungibleHookCall(
                123L, new EvmHookCall(new byte[] {1, 2, 3}, 100000L), FungibleHookType.PRE_TX_ALLOWANCE_HOOK);
        var hookCall2 = new FungibleHookCall(
                456L, new EvmHookCall(new byte[] {4, 5, 6}, 200000L), FungibleHookType.PRE_POST_TX_ALLOWANCE_HOOK);

        tx.addHbarTransferWithHook(accountId1, amount, hookCall1);
        tx.addHbarTransferWithHook(accountId2, amount, hookCall2);

        assertThat(tx.getHbarTransfers()).hasSize(2);
        assertThat(tx.getHbarTransfers().get(accountId1)).isEqualTo(amount);
        assertThat(tx.getHbarTransfers().get(accountId2)).isEqualTo(amount);
    }

    @Test
    void shouldThrowExceptionWhenFrozen() {
        var tx = new TransferTransaction();
        var accountId = new AccountId(0, 0, 1);
        var amount = Hbar.fromTinybars(1000);
        var hookCall = new FungibleHookCall(
                123L, new EvmHookCall(new byte[] {1, 2, 3}, 100000L), FungibleHookType.PRE_TX_ALLOWANCE_HOOK);

        // Set up the transaction properly before freezing
        tx.setTransactionId(TransactionId.withValidStart(AccountId.fromString("0.0.5006"), java.time.Instant.now()));
        tx.setNodeAccountIds(java.util.Arrays.asList(AccountId.fromString("0.0.5005")));

        // Freeze the transaction
        tx.freeze();

        assertThatThrownBy(() -> tx.addHbarTransferWithHook(accountId, amount, hookCall))
                .isInstanceOf(IllegalStateException.class);
    }
}
